esmvalcore/preprocessor/_multimodel.py

```diff
@@ -3,184 +3,65 @@
+}
+
+CONCAT_DIM = 'multi-model'
+
+
+def _resolve_operator(statistic: str):
+    """Find the operator corresponding to the statistic."""
+    statistic = statistic.lower()
+    kwargs = {}
+
+    # special cases
+    if statistic == 'std':
+        logger.warning(
+            "Multicube statistics is aligning its behaviour with iris.analysis"
+            ". Please consider replacing 'std' with 'std_dev' in your code.")
+        statistic = 'std_dev'
+
+    elif re.match(r"^(p\d{1,2})(\.\d*)?$", statistic):
+        # percentiles between p0 and p99.99999...
+        percentile = float(statistic[1:])
+        kwargs['percent'] = percentile
+        statistic = 'percentile'
 
-def _plev_fix(dataset, pl_idx):
-    """Extract valid plev data.
-
-    This function takes care of situations in which certain plevs are
-    completely masked due to unavailable interpolation boundaries.
-    """
-    if np.ma.is_masked(dataset):
-        # keep only the valid plevs
-        if not np.all(dataset.mask[pl_idx]):
-            statj = np.ma.array(dataset[pl_idx], mask=dataset.mask[pl_idx])
-        else:
-            logger.debug('All vals in plev are masked, ignoring.')
-            statj = None
-    else:
-        mask = np.zeros_like(dataset[pl_idx], bool)
-        statj = np.ma.array(dataset[pl_idx], mask=mask)
-
-    return statj
-
-
-def _quantile(data, axis, quantile):
-    """Calculate quantile.
+    try:
+        operator = STATISTIC_MAPPING[statistic]
+    except KeyError as err:
+        raise ValueError(
+            f'Statistic `{statistic}` not supported by multicube statistics. '
+            f'Must be one of {tuple(STATISTIC_MAPPING.keys())}.') from err
 
-    Workaround for calling scipy's mquantiles with arrays of >2
-    dimensions Similar to iris' _percentiles function, see their
-    discussion: https://github.com/SciTools/iris/pull/625
-    """
-    # Ensure that the target axis is the last dimension.
-    data = np.rollaxis(data, axis, start=data.ndim)
-    shape = data.shape[:-1]
-    # Flatten any leading dimensions.
-    if shape:
-        data = data.reshape([np.prod(shape), data.shape[-1]])
-    # Perform the quantile calculation.
-    result = scipy.stats.mstats.mquantiles(data,
-                                           quantile,
-                                           axis=-1,
-                                           alphap=1,
-                                           betap=1)
-    # Ensure to unflatten any leading dimensions.
-    if shape:
-        result = result.reshape(shape)
-    # Check whether to reduce to a scalar result
-    if result.shape == (1, ):
-        result = result[0]
-
-    return result
-
-
-def _compute_statistic(data, statistic_name):
-    """Compute multimodel statistic."""
-    data = np.ma.array(data)
-    statistic = data[0]
-
-    if statistic_name == 'median':
-        statistic_function = np.ma.median
-    elif statistic_name == 'mean':
-        statistic_function = np.ma.mean
```

Comment:
The previous implementation also use `np.ma.mean`, so I still don't get why it's different now?

Reply:
I'm also not sure, but I think maybe it's because there, we specifically allocate the output array as dtype `float32` before filling it with the result of `np.ma.mean`.

https://github.com/ESMValGroup/ESMValCore/blob/3078908caee50d4abc6a1ed4ffbbd994d594f17a/esmvalcore/preprocessor/_multimodel.py#L276-L278

Reply:
Ah yes, that would definitely do that. It is effectively casting to the dtype of the target array, thus would be roughly equivalent to sticking a `.astype(np.float32)` at the array. Should we do that instead?

Reply:
well we need to make sure the data sticks to the rule to be float32 so yeah, I'd argue that's needed

Reply:
In the PR we already have this: https://github.com/ESMValGroup/ESMValCore/blob/223acd7841e8c3e2c2e8c3af0de57c50e80ec02b/esmvalcore/preprocessor/_multimodel.py#L267